File: src/lib/monitor-converter.ts
import { assignmentString, block, convertFromDefinition, map } from "./utils";

const MONITOR = {
  message: (v: any) => assignmentString("message", v),
  name: (v: any) => assignmentString("name", v),
  query: (v: any) => assignmentString("query", v),
  type: (v: any) => assignmentString("type", v),
  options: (v: any) => map(v, (k1, v1) => convertFromDefinition(OPTIONS, k1, v1)),
  id: (_: any) => "",
  tags: (v: any) => assignmentString("tags", v),
  priority: (v: any) => assignmentString("priority", v),
};

const OPTIONS = {
  enable_logs_sample: (v: any) => assignmentString("enable_logs_sample", v),
  escalation_message: (v: any) => assignmentString("escalation_message", v),
  evaluation_delay: (v: any) => assignmentString("evaluation_delay", v),
  force_delete: (v: any) => assignmentString("force_delete", v),
  include_tags: (v: any) => assignmentString("include_tags", v),
  locked: (v: any) => assignmentString("locked", v),
  new_host_delay: (v: any) => assignmentString("new_host_delay", v),
  no_data_timeframe: (v: any) => assignmentString("no_data_timeframe", v),
  notify_audit: (v: any) => assignmentString("notify_audit", v),
  notify_no_data: (v: any) => assignmentString("notify_no_data", v),
  renotify_interval: (v: any) => assignmentString("renotify_interval", v),
  require_full_window: (v: any) => assignmentString("require_full_window", v),
  restricted_roles: (v: any) => assignmentString("restricted_roles", v),
  threshold_windows: (v: any) => block("monitor_threshold_windows", v, assignmentString),
  thresholds: (v: any) => block("monitor_thresholds", v, assignmentString),
  timeout_h: (v: any) => assignmentString("timeout_h", v),
  validate: (v: any) => assignmentString("timeout_h", v),
  groupby_simple_monitor: (v: any) => assignmentString("groupby_simple_monitor", v),
  silenced: (_: any) => "", // 2.23.0 deprecated
};

export function generateTerraformCode(resourceName: string, monitorData: object) {
  let result = "";
  Object.entries(monitorData).forEach(([k, v]) => {
    result += convertFromDefinition(MONITOR, k, v);
  });
  return `resource "datadog_monitor" "${resourceName}" {${result}\n}`;
}

File: cmd/monitor-converter.go
package main

import (
	"fmt"
	"strings"
)

// MONITOR and OPTIONS definitions
var MONITOR = map[string]stringFunc{
	"message": stringGen("message"),
	"name":    stringGen("name"),
	"query":   stringGen("query"),
	"type":    stringGen("type"),
	"options": func(v any) string {
		return mapContents(v.(map[string]any), func(k1 string, v1 any) string { return convertFromDefinition(OPTIONS, k1, v1) })
	},
	"id":               blankGen,
	"tags":             stringGen("tags"),
	"priority":         stringGen("priority"),
	"restricted_roles": stringGen("restricted_roles"),
}

var OPTIONS = map[string]stringFunc{
	"enable_logs_sample":  stringGen("enable_logs_sample"),
	"escalation_message":  stringGen("escalation_message"),
	"evaluation_delay":    stringGen("evaluation_delay"),
	"force_delete":        stringGen("force_delete"),
	"include_tags":        stringGen("include_tags"),
	"locked":              stringGen("locked"),
	"new_host_delay":      stringGen("new_host_delay"),
	"no_data_timeframe":   stringGen("no_data_timeframe"),
	"notify_audit":        stringGen("notify_audit"),
	"notify_no_data":      stringGen("notify_no_data"),
	"renotify_interval":   stringGen("renotify_interval"),
	"require_full_window": stringGen("require_full_window"),
	"restricted_roles":    stringGen("restricted_roles"),
	"threshold_windows": func(v any) string {
		return block("monitor_threshold_windows", v.(map[string]any), assignmentString)
	},
	"thresholds": func(v any) string {
		return block("monitor_thresholds", v.(map[string]any), assignmentString)
	},
	"timeout_h":              stringGen("timeout_h"),
	"validate":               stringGen("timeout_h"),
	"groupby_simple_monitor": stringGen("groupby_simple_monitor"),
	"silenced":               blankGen, // Deprecated
	"new_group_delay":        stringGen("new_group_delay"),
	"renotify_statuses":      stringGen("renotify_statuses"),
	"on_missing_data":        stringGen("on_missing_data"),
}

func generateMonitorTerraformCode(resourceName string, monitorData map[string]any) string {
	var result strings.Builder
	for k, v := range monitorData {
		result.WriteString(convertFromDefinition(MONITOR, k, v))
	}
	return fmt.Sprintf("resource \"datadog_monitor\" \"%s\" {%s\n}", resourceName, result.String())
}
